File: app/src/main/java/com/arejas/dashboardofthings/domain/entities/result/Resource.java
package com.arejas.dashboardofthings.domain.entities.result;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Objects;

/*
 * Class used for representing a request to database or network and its result.
 */
public class Resource<T> {
    @NonNull
    private Status status;
    @Nullable
    private T data;
    @Nullable
    private Throwable error;

    private Resource(@NonNull Status status, @Nullable T data,
                     @Nullable Throwable error) {
        this.status = status;
        this.data = data;
        this.error = error;
    }

    public static <T> Resource<T> success(@Nullable T data) {
        return new Resource<>(Status.SUCCESS, data, null);
    }

    public static <T> Resource<T> error(Throwable error, @Nullable T data) {
        return new Resource<>(Status.ERROR, data, error);
    }

    public static <T> Resource<T> loading(@Nullable T data) {
        return new Resource<>(Status.LOADING, data, null);
    }

    @NonNull
    public Status getStatus() {
        return status;
    }

    @Nullable
    public T getData() {
        return data;
    }

    @Nullable
    public Throwable getError() {
        return error;
    }

    public void setStatus(@NonNull Status status) {
        this.status = Objects.requireNonNull(status);
    }

    public void setData(@Nullable T data) {
        this.data = data;
    }

    public void setError(@Nullable Throwable error) {
        this.error = error;
    }

    public enum Status { SUCCESS, ERROR, LOADING }
}

File: app/src/main/java/com/arejas/dashboardofthings/domain/entities/result/Resource.kt
package com.arejas.dashboardofthings.domain.entities.result

import java.util.Objects

/*
 * Class used for representing a request to database or network and its result.
 */
class Resource<T> private constructor(
    var status: Status, var data: T?,
    var error: Throwable?
) {

    enum class Status {
        SUCCESS, ERROR, LOADING
    }

    companion object {

        fun <T> success(data: T?): Resource<T> {
            return Resource(Status.SUCCESS, data, null)
        }

        fun <T> error(error: Throwable, data: T?): Resource<T> {
            return Resource(Status.ERROR, data, error)
        }

        fun <T> loading(data: T?): Resource<T> {
            return Resource(Status.LOADING, data, null)
        }
    }
}
